### ISSUE TITLE
compute_gradient w.r.t flow-velocity for advection equation: Errors in taylor_test and compute_gradient

### ISSUE THREAD

**Issue by BeccyB** [dolfin-adjoint/pyadjoint#42]
Hi,

I am currently facing two errors and truly have no idea what I am doing wrong :thinking: Just started to work with dolfin-adjoint and fenics a few weeks ago and maybe I am overlooking something obvious. Any help or ideas on how to debug or fix one of the errors would be very much appreciated :slightly_smiling_face:

The code below is adapted from the advection-diffusion  undocumented demos to be a time dependent advection problem. With the initial condition u_0 = sin(2pi x) the exact solution if beta = const is u(x,t) = sin(2pi(x-beta_x t) is also computed correctly by the code.

Error 1:

I would like to add an optimization routine w.r.t to the advection velocity in x-direction beta = [beta_x, beta_y]^T = [0.5, 0]^T. (This is the simplest example just for testing purposes). 

The derivative dJdbeta_x and the reducedFunctional is computed with out giving any error. However, if I now want to test the results with the taylor_test I get the following error message:

    ~/anaconda3/envs/dolfin_master/lib/python3.8/site-packages/dolfin/fem/solving.py in _extract_args(*args, **kwargs)
        320     for kwarg in kwargs.keys():
        321         if kwarg not in valid_kwargs:
    --> 322             raise RuntimeError("Solve variational problem. Illegal keyword argument \'{}\'.".format(kwarg))
        323 
        324     # Extract equation

    RuntimeError: Solve variational problem. Illegal keyword argument 'function'.

Error 2:
Eventually, I would also like to define the velocity as an Expression so that I can have beta dependent on x[0] and/or x[1]. 

UPDATED: I found out that when using an expression, one has to supply the derivative and dependencies. 

```
beta_x = Expression("0.5", degree=1) 
beta_y = Expression("0.0", degree=1)
beta = Expression(("beta_x","beta_y"), degree=2, beta_x = beta_x, beta_y=beta_y)
beta.dependencies = [beta_x, beta_y]
beta.user_defined_derivatives = {beta_x: Expression("0",degree=1), beta_y : Expression("0",degree=1)}
```
But even if I supply these I get the same error:

```
NotImplementedError                       Traceback (most recent call last)
<ipython-input-13-91f6fd6fb64c> in <module>
     78 
     79 J = assemble(inner(u_old, u_old)*dx)
---> 80 dJdbeta0 = compute_gradient(J, Control(beta_x))
....
~/anaconda3/envs/dolfin_master/lib/python3.8/site-packages/pyadjoint/overloaded_type.py in _ad_convert_type(self, value, options)
    133 
    134         """
--> 135         raise NotImplementedError
    136 
    137     def _ad_create_checkpoint(self):

```

This is a shorted version of the code to reproduce the error message:

    from dolfin import *
    from fenics_adjoint import *
    import matplotlib.pyplot as plt
    import numpy as np

    # adapted code from
    # """ Steady state advection-diffusion equation,
    # discontinuous formulation using full upwinding.
    # Implemented in python from cpp demo by Johan Hake.
    # """

    m_size = 64
    mesh = UnitSquareMesh(m_size, m_size)
    parameters["ghost_mode"] = "shared_facet"

    delta_t = 0.01
    end_time = 1
    time_steps = int(end_time/delta_t)

    V_dg = FunctionSpace(mesh, "DG", 1)
    
    beta_x, beta_y = Constant(0.5), Constant(0.0) # Causes Error in line taylor_test
    
    # Eventually I would like beta not be constant but dependent on x/y 
    # Causes not implemented error in compute_gradient
    # beta_x = Expression("0.5",degree=1)
    # beta_y = Expression("0.0", degree=1)
    # beta = Expression(("beta_x","beta_y"), degree=2, beta_x = beta_x, beta_y=beta_y)
    # beta.dependencies = [beta_x, beta_y]
    # beta.user_defined_derivatives = {beta_x: Expression("0",degree=1), beta_y : Expression("0",degree=1)}

    beta = as_vector([beta_x, beta_y])

    walls   = 'near(x[1], 0) || near(x[1], 1) || near(x[0], 1) || near(x[0], 0)'
    u_exact = Expression("sin(2*pi*(x[0] - beta_x*t))", degree=3, t=0, beta_x=beta_x)
    bcs     = DirichletBC(V_dg, u_exact, walls, method='geometric') 

    phi,v = TrialFunction(V_dg), TestFunction(V_dg)
    u_new, u_old = Function(V_dg), Function(V_dg)

    f = Constant(0.0)
    u0 = Expression("sin(2*pi*x[0])", degree=1, t=0)

    n = FacetNormal(mesh)
    h = CellDiameter(mesh)
    h_avg = (h('+') + h('-'))/2
    un = (dot(beta, n) + abs(dot(beta, n)))/2.0

    a_int = dot(grad(v), - beta*phi)*dx
    a_vel = dot(jump(v), un('+')*phi('+') - un('-')*phi('-'))*dS
    a_boundary = dot(v, u_exact*phi)*ds
    a = a_vel + a_int + a_boundary + Constant(1/delta_t)*phi*v*dx
    L = v*f*dx + Constant(1/delta_t)*u_old*v*dx

    project(u0, V_dg, function=u_old)

    plot(u_old)
    plt.title("Initial condition at t=0")
    plt.show()

    t = 0

    for n in range(time_steps):
            t += delta_t
            u_exact.t = t

            solve(a==L, u_new, bcs)
            u_old.assign(u_new)

    plot(u_old)
    plt.title("Solution at t=1")
    plt.show()

    # -----------------------------------
    # Gradient computation 
    # -----------------------------------
    J = assemble(inner(u_old, u_old)*dx)
    dJdbeta_x = compute_gradient(J, Control(beta_x))
    Jhat = ReducedFunctional(J, Control(beta_x))
    h = Constant(0.01)
    conv_rate = taylor_test(Jhat, beta_x, h)

I use dolfin-adjoint 2019.1.0

**Comment by BeccyB** [comment]
Hi everyone,

finally found what was causing the RuntimeError in `taylor_test`. I created the function u_old but initialize it with u0 **after** defining the problem a, L.  If I move the project(u0, V_dg, function=u_old) before the definition of a, L, then I do not get any errors.  

Could someone perhaps explain why the IC must be defined **before** a,L is defined? As far as I can tell for FEniCS the position of the projection of the initial condition does not change the solution in anyway. It seems to have an influence to doflin-adjoint or am I misunderstanding something here?

Many thanks in advance!

**Comment by sebastkm** [comment]
Hi,

The first error is due to the `function` keyword argument in the project function for your IC on the line `project(u0, V_dg, function=u_old)`. I've made a fix for that in PR #46.

The second error when you instead use expressions is caused by `beta_x` now being an `Expression` and dolfin-adjoint is not able to differentiate w.r.t an expression, it can only differentiate through it (in other words, Expressions can't be controls/leaf nodes in the computational graph, but you can differentiate w.r.t parameters used as inputs to the expression).

I'm not sure why moving the projection to before a and L are defined circumvents the problem, though. I can't reproduce that workaround with your example code.

**Comment by BeccyB** [comment]
Hi sebastkm,

thank you for fixing the project function. This seemed to be the real underlying issue.
